### PR TITLE
Improving JSON Schema Validation Errors

### PR DESCRIPTION
**Pull request by robons**
Addresses tickets  #512 And #515.

Gives deeper and potentially easier to debug validation errors like the following example:

```
2022-09-15 17:33:36,060 - csvcubed.cli.build - WARNING - Schema Validation Error: $.columns.Dim-1 - Unable to identify {'type': 'dimension', 'from_existing': 'https://example.com/some-existing-dimension'}

    If you meant to declare 'This column represents a dimension of the cube', then:
        $.from_existing - 'https://example.com/some-existing-dimension' is not recognised by csvcubed.

    If you meant to declare 'This column represents an attribute (with literal values) of the cube', then:
        $.type - 'attribute' was expected
        $.from_existing - 'https://example.com/some-existing-dimension' is not recognised by csvcubed.

    If you meant to declare 'This column represents an attribute (with resource values) of the cube', then:
        $.type - 'attribute' was expected
        $.from_existing - 'https://example.com/some-existing-dimension' is not recognised by csvcubed.

    If you meant to declare 'This column represents the phenomenon which has been measured', then:
        - Additional properties are not allowed ('from_existing' was unexpected)
        $.type - 'measures' was expected

    If you meant to declare 'This column holds observed/recorded values', then:
        - Additional properties are not allowed ('from_existing' was unexpected)
        $.type - 'observations' was expected

    If you meant to declare 'This column represents the unit of measure in which observations were recorded', then:
        $.type - 'units' was expected
        - Additional properties are not allowed ('from_existing' was unexpected)

    If you meant to declare 'This column represents the unit of measure in which observations were recorded', then:
        $.type - 'units' was expected
        - Additional properties are not allowed ('from_existing' was unexpected) (build.py:105)
```

As usual, don't be concerned about checking the VCRpy cassetted (the YAML files).

### CODE REVIEW

File: tests/unit/readers/cubeconfig/v1_0/test_cube_from_convention.py [hunk 1]
```diff
@@ -206,7 +206,7 @@ def test_unsupported_col_definition_exception():
         deserialiser = get_deserialiser(SCHEMA_PATH_FILE, 3)
 
         with pytest.raises(UnsupportedColumnDefinitionException) as msg:
-            _, _, _ = deserialiser(data_file_path, config_file_path)
```
Comment (GDonRanasinghe) [inline tests/unit/readers/cubeconfig/v1_0/test_cube_from_convention.py:209]: I dont know why I did this!

File: tests/unit/utils/test_text.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    assert len(truncated_message) == 11  # We expect an extra char for the ellipsis
+    assert truncated_message == "This messa…"
+
+
+def test_text_not_needing_truncation():
+    """Ensure that text truncation doesn't affect a message shorter than the limit."""
+    truncated_message = truncate("A short message", 20)
+    assert len(truncated_message) == len("A short message")
+    assert truncated_message == "A short message"
+
```
Comment (GDonRanasinghe) [inline tests/unit/utils/test_text.py:18]: What will happen if `truncate("", 10)`? I believe as same as in `test_text_not_needing_truncation`?

**Review by GDonRanasinghe — APPROVED** [review]
This looks good to me. I have added a couple of comments for you to check before merging.

File: src/csvcubed/cli/build.py [hunk 3]
```diff
@@ -46,18 +47,15 @@ def build(
...
-            validation_errors_dict = [
-                e.as_json_dict()
-                if isinstance(e, DataClassBase)
-                else dataclasses.asdict(e)
-                for e in validation_errors
-            ]
-            all_errors = validation_errors_dict + [
-                e.message for e in json_schema_validation_errors
+            all_errors: List[ValidationError] = (
+                validation_errors + json_schema_validation_errors  # type: ignore
```
[the remark this answers is not in the document]
Comment (robons) [reply src/csvcubed/cli/build.py:51]: Yes, it started to complain that it wasn't possible to concatenate two lists with different types inside them.

File: tests/unit/utils/test_text.py (hunk 2 above)
Comment (robons) [reply tests/unit/utils/test_text.py:18]: Yes, the same behaviour will happen.

File: src/csvcubed/cli/build.py [hunk 4]
```diff
@@ -79,20 +77,40 @@ def build(
...
+def _validation_error_to_display_json_dict(error: ValidationError) -> dict:
+    dict_value: dict
+    if isinstance(error, DataClassBase):
+        dict_value = error.as_json_dict()
+    else:
+        dict_value = dataclasses.asdict(error)
+
+    if isinstance(error, JsonSchemaValidationError):
+        # Remove schema from output since it's verbose and unhelpful for users.
+        del dict_value["schema"]
```
Comment (robons) [inline src/csvcubed/cli/build.py:89]: todo: Need to do a bit of recursion here to sort out child errors.

**Pull request merged**